components/builder-graph/src/package_graph.rs

```diff
@@ -0,0 +1,795 @@
+            static ref NO_DEPS_RE: Regex =
+                Regex::new(r"no (build|runtime) dependencies or undefined").unwrap();
+            static ref GEN_DEP_RE: Regex =
+                Regex::new(r"^\s*\* __(?P<dtype>Build)?\s*Dependencies__: `?(?P<deps>[^`]*)`?\s*$")
+                    .unwrap();
+        }
+
+        let package = self.package.as_ref().unwrap();
+        let mut found_deps = false;
+        let mut found_bdeps = false;
+
+        // investigate RegexSet usage here instead of looping over lines
+        for line in package.manifest.lines() {
+            if let Some(cap) = GEN_DEP_RE.captures(line) {
+                if verbose {
+                    println!("{} matched line {}", package.ident.0, line);
+                }
+                let deplist = cap.name("deps").unwrap().as_str();
+                if verbose {
+                    println!("{} extracted deps {}", package.ident.0, deplist);
+                }
+                // Maybe match against regex 'no (build|runtime) dependencies or undefined'
+                let mut deps_as_ident = if !deplist.contains("dependencies or undefined") {
+                    let deps_conv: herror::Result<Vec<PackageIdent>> = deplist
+                        .split_whitespace()
+                        .map(PackageIdent::from_str)
+                        .collect();
+
+                    deps_conv.unwrap_or_else(|e| {
+                        println!("{} ill formed deps {:?}", package.ident.0, e);
+                        Vec::new()
+                    })
+                } else {
+                    Vec::new()
+                };
+
+                let typeflag;
+                if let Some(_deptype) = cap.name("dtype") {
+                    typeflag = "B";
+                    found_bdeps = true;
+                    if verbose {
+                        println!(
+                            "{} {}: {:?}",
+                            package.ident.0,
+                            typeflag,
+                            join_idents(", ", &deps_as_ident)
+                        )
+                    };
+                    self.plan_bdeps.append(&mut deps_as_ident);
+                } else {
+                    typeflag = "R";
+                    found_deps = true;
+                    if verbose {
+                        println!(
+                            "{} {}: {:?}",
+                            package.ident.0,
+                            typeflag,
+                            join_idents(", ", &deps_as_ident)
+                        );
+                    };
+                    self.plan_deps.append(&mut deps_as_ident);
+                }
+            }
+
+            // early out; manifests can be large...
+            if found_deps && found_bdeps {
+                break;
+            }
+        }
+
+        if !(found_deps && found_bdeps) {
+            // Every package should have deps; if not we need to recheck our assumptions
```

Comment:
I wonder about this. We do have the ability for users to package software using their underlying system deps (though it is not encouraged and to my knowledge we've refrained from heavily documenting the behavior.

Reply:
Because the studio provides a base set of tooling, in at least one package we've gotten away without declaring build deps:

https://github.com/habitat-sh/core-plans/blob/master/cacerts/plan.sh#L10-L11

I could imagine other such "data" type packages having no deps. Not sure if that applies here.

Reply:
Yeah, that was an overly optimistic view of the world. Not every package has deps. There are a few classes this falls into:
1) True 'no deps' packages. core/cacerts is a good example
2) User packages statically built using system dependencies. While not 'pure' habitat packages, this is a supported use case and common in windows.
3) Bad packages... There are a lot of packages that don't list deps in their manifest, but have them in their plan file. Or don't list them in their plan file, but use them some how.

There are a lot of bad packages, but unfortunately there are enough legit no-deps packages that we can't just flag them as bad.